2020-02-Spring/2_linked_lists/README.md

```diff
@@ -126,31 +147,117 @@ Go to [Top](#top)
 <!-- Don't remove -->
 <a name="s2"/>
 
-### 2. SOLUTION 2 TODO :bug:
-
-Source: TODO :bug:
+### 2. Remove Zero Sum Consecutive Nodes from Linked List
 
 #### Naive/Simple Solution
+The idea is to locate the sublist of nodes that sum to 0 and de-list them.<br>
+- Starting at the edging node, find lower-bound and upper-bound node of the sublist that sum to 0.
+- if found, de-list nodes from [LBound, UBound], including by re-link PreLBound->next = UBound->next;
+- If not found, move forward and check the next node.
+For every i-th node of N nodes, the algorithm iterates from node i-th to the last node for the total of N*(N-i) times.
+- This solution has time complexity of O(N<sup>2</sup>).
 
 <details>
 <summary>Click to see solution</summary>
 
-TODO put your solution here :bug:
+```C++
+ListNode* removeZeroSumSublists(ListNode* head) {
+    ListNode *Edge, *PreLBound, *UBound;
+    // create edging node
+    Edge= new ListNode(0);
+    Edge->next = head;
+    PreLBound = Edge;
+    UBound = head;
+    int Sum=0;
+    while(UBound)
+    {
+        Sum = Sum + UBound->val;
+
+        // if found zero sum range [LBound,UBound]
+        // where LBound == PreLBound->Next;
+        // then re-link from Prev to UBound->Next
+        if(Sum == 0) {
+            PreLBound->next = UBound->next;
+        }
+        UBound = UBound->next;
+        // no zero-sum found in the range [Prev, End]
+        // continue checking from [Prev->next, End]
+        if(UBound == NULL)
+        {
+            PreLBound = PreLBound->next;
+            // end of list
+            if(PreLBound == NULL) break;
+            UBound = PreLBound->next;
+            Sum=0;
+        }
+    }
+    // returning the result without the edging node
+    return Edge->next;
+}
+```
 
 </details>
 
 #### Optimal Solution
+One approach is to calculate the accumulated sum that is defined by Sum<sub>i</sub> = Sum <sub>i - 1</sub> + List <sub>i</sub> -> val. The accumulated Sum is tracked by using the hash table SumMap.
+- if Sum is found in SumMap (accumulated Sum is repeated), this means there is a zero-sum range existing between the two nodes, then we remove all nodes between those two(s) and continue scanning until no zero-sum range found.
+- if Sum is not repeated, this means there is NO zero-sum range existing between the two nodes, therefore, we will record the accumulated sum and examine the next node. <br>
+- Instead of going through from (N-i) nodes for every i-th node explained in the previous solution, this solution uses a hash table to record the node before the zero-sum range and only goes through the list once.
+- Time complexity of this solution is O(N). <br>
```

Comment:
A few suggestions:

1. I recommend using monospace for all code-like snippets you're including here. The subscripts are neat, but inline code snippets draws a better distinction for variable names.
2. No `<br>` elements are required in an unordered list (which is what the `- ` prefix does before each line.
3. This is a good solution, which optimizes for time complexity, **but takes your space complexity from `O(1)` to `O(N)`** because you are allocating additional space, potentially one extra map entry for every element in your list. You should mention this tradeoff.